File: containersMostWater.py
"""
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
"""
class Solution:
    def maxArea(self, height: List[int]) -> int:
        
        first = 0
        last = len(height) - 1
        maxArea = float("-inf")
        
        while(first <= last):
            
            # Area of this tank
            heightOfTank = min(height[first], height[last])
            widthOfTank = last - first
            areaOfTank = heightOfTank * widthOfTank
            
            # Check if this tank is max tank
            maxArea = max(maxArea, areaOfTank)
            
            # Move the last pointer if last < first
            if(height[first] > height[last]):
                last -= 1
            else:
                first += 1
                
        return maxArea

File: containers_most_water.cpp
// Time Complexity : O(n)
// Space Complexity :  O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this

class Solution {
public:
    int maxArea(vector<int>& height) {
        
        int len = height.size();
        int low = 0;
        int high = len-1;
        
        int max_area = INT_MIN;
        int temp_area;
        
        while(low < high)
        {
            temp_area = min(height[low], height[high]) * (high - low);
            
            if(temp_area > max_area)
                max_area = temp_area;
            
            if(height[low] < height[high])
                low++;
            else
                high--;
        }
        
        return max_area;
    }
};